### PR TITLE
feat(provider/ecs): ECR Image Provider

### PR DESCRIPTION
**Pull request by dkirillov**
Adds an `EcrImageProvider` which is responsible for retrieving ECR image during the findImage stage. 

@cfieber @ajordens @robzienert @BrunoCarrier

### CODE REVIEW

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    String accountId = extractAwsAccountId(url);
+    String repository = extractEcrRepositoryName(url);
+    String identifier = extractEcrIdentifier(repository, url);
+    boolean isTag = !(identifier.startsWith("sha256:") && identifier.length() == ("sha256:".length() + 64));
+    String region = extractAwsRegion(url);
+
+    NetflixAmazonCredentials credentials = getCredentials(accountId);
+
+    if (!isValidRegion(credentials, region)) {
+      throw new Error("The repository URI provided does not belong to a region that the credentials have access to or is not a valid region.");
```
Comment (robzienert) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java]: `s/or is not a valid region/or the region is not valid`.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+  private String extractAwsRegion(String imageUrl) {
+    return extractString(REGION_PATTERN, imageUrl, 0,
+      "The repository URI provided does not contain a proper region.");
+  }
+
+  private String extractString(Pattern pattern, String imageUrl, int group, String error) {
+    Matcher matcher = pattern.matcher(imageUrl);
+    if (!matcher.find()) {
+      throw new Error(error);
```
Comment (robzienert) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java]: What is `Error`? Do you think we should have different exceptions based on the condition? I suspect if we need to do error handling, we wouldn't want to have a `Error`, as that'd be the same anti-pattern of doing a catch on `Exception` or `Throwable` -- too broad.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+  @Override
+  public List<EcsDockerImage> findImage(String url) {
+    // HTTP(S) part is not needed.
+    url = url.replace("http://", "").replace("https://", "");
+
+    String accountId = extractAwsAccountId(url);
+    String repository = extractEcrRepositoryName(url);
+    String identifier = extractEcrIdentifier(repository, url);
+    boolean isTag = !(identifier.startsWith("sha256:") && identifier.length() == ("sha256:".length() + 64));
```
Comment (robzienert) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java]: The digest is always a sha256?

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java (hunk 1 above)
Comment (dkirillov) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java]: Done.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java (hunk 3 above)
Comment (dkirillov) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java]: Yes, the digest is always a sha256.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java (hunk 2 above)
Comment (dkirillov) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java]: Changed it to `IllegalArgumentException` - it comes down to the passed in (ecr) `url` being improper.

**Pull request merged**